libs/autotagging-commons/src/main/java/org/opensearch/rule/RuleQueryBuilder.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rule;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.rule.action.GetRuleRequest;
+
+/**
+ * This interface is responsible for creating query objects which storage layer can use
+ * to query the backend
+ * @param <T>
+ */
+@ExperimentalApi
+public interface RuleQueryBuilder<T> {
+    /**
+     * This method translates the {@link GetRuleRequest} to a storage engine specific  query object
+     * @param request
+     * @return
+     */
+    T buildQuery(GetRuleRequest request);
```

Comment:
Why tie this to GetRuleRequest? I guess even your UpdateRuleRequest can also reuse this to return a query object? 

Would be better to use a common class here which could be used by both if possible. If not possible, rename this method to convey this is meant for `GetRuleRequest` only.

Also considering you are converting one object(request) into another(query), this might fit into a "mapper" pattern rather than builder.

Reply:
Good point! let me think a better way to organise this

Reply:
We can introduce use case specific methods for the query building part and reuse what can be reused inside the implementation